transmart-core-db/grails-app/services/org/transmartproject/db/clinical/CrossTableService.groovy

```diff
@@ -0,0 +1,66 @@
+package org.transmartproject.db.clinical
+
+import groovy.transform.CompileStatic
+import org.springframework.beans.factory.annotation.Autowired
+import org.transmartproject.core.exceptions.AccessDeniedException
+import org.transmartproject.core.multidimquery.CrossTable
+import org.transmartproject.core.multidimquery.CrossTableResource
+import org.transmartproject.core.multidimquery.PatientSetResource
+import org.transmartproject.core.multidimquery.query.Combination
+import org.transmartproject.core.multidimquery.query.Constraint
+import org.transmartproject.core.multidimquery.query.Operator
+import org.transmartproject.core.multidimquery.query.PatientSetConstraint
+import org.transmartproject.core.users.User
+import org.transmartproject.db.multidimquery.CrossTableImpl
+
+@CompileStatic
+class CrossTableService extends AbstractDataResourceService implements CrossTableResource {
+
+    @Autowired
+    PatientSetResource patientSetResource
+
+    @Autowired
+    AggregateDataService aggregateDataService
+
+    @Override
+    CrossTable retrieveCrossTable(List<Constraint> rowConstraints, List<Constraint> columnConstraints,
+                                           Long patientSetId, User user) {
+
+        Constraint patientSetConstraint = new PatientSetConstraint(patientSetId)
+        return buildCrossTable(rowConstraints, columnConstraints, patientSetConstraint, user)
+    }
+
+    private CrossTable buildCrossTable(List<Constraint> rowConstraints,
+                                                           List<Constraint> columnConstraints,
+                                                           Constraint patientSetConstraint,
+                                                           User user) {
+        log.info "Building a cross table..."
+        List rows = []
+        for (Constraint rowConstraint : rowConstraints) {
+            def counts = []
+            for (Constraint columnConstraint : columnConstraints) {
+                counts.add(getCrossTableCell(rowConstraint, columnConstraint, patientSetConstraint, user))
+            }
+            rows.add(new CrossTableImpl.CrossTableRowImpl(counts))
+        }
+        new CrossTableImpl(rows)
+    }
+
+    private Long getCrossTableCell(Constraint rowConstraint,
+                                   Constraint columnConstraint,
+                                   Constraint patientSetConstraint,
+                                   User user) {
+        def constraints = [rowConstraint, columnConstraint, patientSetConstraint] as List<Constraint>
+        for(Constraint constraint in constraints) {
+            try {
+                checkAccess(constraint, user)
+            } catch(AccessDeniedException e) {
+                log.warn e.message
+                return 0
+            }
+        }
+        Constraint crossConstraint = new Combination(Operator.AND, constraints)
+        return aggregateDataService.counts(crossConstraint, user).patientCount
```

Comment:
This method does most of the job to create the cross table, but I am not sure if that is all that is required for the cross table in terms of creating/reusing/cashing patient sets/bitsets. @gijskant is this enough?

Reply:
For using bitsets and reusing patient sets this is not enough, but that doesn't matter for now. This can be our first (reference) implementation that we can later optimise for large datasets easily, but saving a patient set for each of the (row and column) constraints.
I do think that instead of a combination it should be a multiplesubselect. This works fine as long as each of the row and column constraints are patient set constraints or subselects, but if they are value or concept constraints you will get empty results.

Reply:
@gijskant @forus I wanted to replace Combination constraint with MultipleSubSelectionsConstraint (as suggested by Gijs), but [this line of AggregateDataService|https://github.com/thehyve/transmart-core/blob/dev/transmart-core-db/grails-app/services/org/transmartproject/db/clinical/AggregateDataService.groovy#L206] gives me UnsupportedOperationException pointing to ConstraintSerialiser.build(ConstraintSerialiser.groovy:129). MultipleSubSelectionsConstraint was changed to be unsupported in ConstraintSerialiser in [b673b06e|https://github.com/thehyve/transmart-core/commit/b673b06e02bacb59d3a530375a40e20a1bdbc8dc] because it should be for internal use only, but in this case I want to use it internally. Should I change AggregateDataService or ConstraintSerialiser then?

Reply:
Or use `new AndConstraint([rowConstraint, columnConstraint, subjectConstraint].collect { new SubSelection(it) })])` or something similar. It is equivalent to the multiplesubselections.

Reply:
So, wrapping each of the constraints in a subselect is sufficient. Multiple subselections is just a more compact way of writing that, with the limitation that you cannot serialise it.

Reply:
ah ok, thanks! I will replace MultipleSubSelectionsConstraint with a mix of And/SubSelection constraints then